language_server/language_server.cpp

```diff
@@ -0,0 +1,131 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "language_server/language_server.h"
+
+#include "third_party/clangd/Protocol.h"
+#include "toolchain/diagnostics/null_diagnostics.h"
+#include "toolchain/lexer/tokenized_buffer.h"
+#include "toolchain/parser/parse_node_kind.h"
+#include "toolchain/parser/parse_tree.h"
+#include "toolchain/source/source_buffer.h"
+
+namespace Carbon::LS {
+
+namespace {
+template <typename T>
+auto ParseJSON(llvm::json::Value& v, T& parsed) -> bool {
+  llvm::json::Path::Root root_path;
+  return clang::clangd::fromJSON(v, parsed, root_path);
+}
+}  // namespace
+
+auto LanguageServer::onNotify(llvm::StringRef method, llvm::json::Value value)
+    -> bool {
+  if (method == "textDocument/didOpen") {
+    clang::clangd::DidOpenTextDocumentParams params;
+    if (!ParseJSON(value, params)) {
+      return false;
+    }
+    files_.emplace(params.textDocument.uri.file(), params.textDocument.text);
+  }
+  if (method == "textDocument/didChange") {
```

Comment:
This is an alternative, not a case where both `if` branches could be taken.

```suggestion
  } else if (method == "textDocument/didChange") {
```